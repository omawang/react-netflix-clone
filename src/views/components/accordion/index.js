import React, { createContext, useContext, useState } from 'react';
import * as S from './styles/accordion';

const ToggleContext = createContext();

export default function Accordion({ children, ...restProps }) {
  return (
    <S.Container {...restProps}>
      <S.Inner>{children}</S.Inner>
    </S.Container>
  );
}

Accordion.Title = function AccordionTitle({ children, ...restProps }) {
  return <S.Title {...restProps}>{children}</S.Title>;
};

Accordion.Item = function AccordionItem({ children, ...restProps }) {
  const [toggleShow, setToggleShow] = useState(false);

  return (
    <ToggleContext.Provider value={{ toggleShow, setToggleShow }}>
      <S.Item {...restProps}>{children}</S.Item>
    </ToggleContext.Provider>
  );
};

Accordion.Header = function AccordionHeader({ children, ...restProps }) {
  const { toggleShow, setToggleShow } = useContext(ToggleContext);

  return (
    <S.Header
      onClick={() => setToggleShow((toggleShow) => !toggleShow)}
      {...restProps}
    >
      {children}
      {toggleShow ? (
        <img src="/images/icons/close-slim.png" alt="Close" />
      ) : (
        <img src="/images/icons/add.png" alt="Open" />
      )}
    </S.Header>
  );
};

Accordion.Body = function AccordionBody({ children, ...restProps }) {
  const { toggleShow } = useContext(ToggleContext);

  return toggleShow ? <S.Body {...restProps}>{children}</S.Body> : null;
};
