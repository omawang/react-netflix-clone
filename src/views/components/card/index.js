import React, { createContext, useContext, useState } from 'react';

import * as S from './styles/card';

export const FeatureContext = createContext();

export default function Card({ children, ...restProps }) {
  const [showFeature, setShowFeature] = useState(false);
  const [itemFeature, setItemFeature] = useState({});

  return (
    <FeatureContext.Provider
      value={{ showFeature, setShowFeature, itemFeature, setItemFeature }}
    >
      <S.Container {...restProps}>{children}</S.Container>
    </FeatureContext.Provider>
  );
}

Card.Group = function CardGroup({ children, ...restProps }) {
  return <S.Group {...restProps}>{children}</S.Group>;
};

Card.Title = function CardTitle({ children, ...restProps }) {
  return <S.Title {...restProps}>{children}</S.Title>;
};

Card.Entities = function CardEntities({ children, ...restProps }) {
  return <S.Entities {...restProps}>{children}</S.Entities>;
};

Card.Item = function CardItem({ item, children, ...restProps }) {
  const { setShowFeature, setItemFeature } = useContext(FeatureContext);

  return (
    <S.Item
      onClick={() => {
        setItemFeature(item);
        setShowFeature(true);
      }}
      {...restProps}
    >
      {children}
    </S.Item>
  );
};

Card.Image = function CardImage({ src, ...restProps }) {
  return <S.Image src={src} {...restProps} />;
};

Card.Meta = function CardMeta({ children, ...restProps }) {
  return <S.Meta {...restProps}>{children}</S.Meta>;
};

Card.SubTitle = function CardSubTitle({ children, ...restProps }) {
  return <S.SubTitle {...restProps}>{children}</S.SubTitle>;
};

Card.Text = function CardText({ children, ...restProps }) {
  return <S.Text {...restProps}>{children}</S.Text>;
};

Card.Feature = function CardFeature({ category, children, ...restProps }) {
  const { showFeature, itemFeature, setShowFeature } = useContext(
    FeatureContext
  );

  return showFeature ? (
    <S.Feature
      src={`/images/${category}/${itemFeature.genre}/${itemFeature.slug}/large.jpg`}
      {...restProps}
    >
      <S.FeatureContent>
        <S.FeatureTitle>{itemFeature.title}</S.FeatureTitle>
        <S.FeatureText>{itemFeature.description}</S.FeatureText>
        <S.FeatureClose onClick={() => setShowFeature(false)}>
          <img src="/images/icons/close.png" alt="Close" />
        </S.FeatureClose>

        <S.Group margin="30px 0" flexDirection="row" alignItems="center">
          <S.Maturity rating={itemFeature.maturity}>
            {itemFeature.maturity < 12 ? 'PG' : itemFeature.maturity}
          </S.Maturity>
          <S.FeatureText fontWeight="bold">
            {itemFeature.genre.charAt(0).toUpperCase() +
              itemFeature.genre.slice(1)}
          </S.FeatureText>
        </S.Group>
        {children}
      </S.FeatureContent>
    </S.Feature>
  ) : null;
};
