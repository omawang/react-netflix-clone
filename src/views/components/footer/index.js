import React from 'react';
import * as S from './styles/footer';

export default function Footer({ children, ...restProps }) {
  return <S.Container {...restProps}>{children}</S.Container>;
}

Footer.Row = function FooterRow({ children, ...restProps }) {
  return <S.Row {...restProps}>{children}</S.Row>;
};

Footer.Column = function FooterColumn({ children, ...restProps }) {
  return <S.Column {...restProps}>{children}</S.Column>;
};

Footer.Link = function FooterLink({ children, ...restProps }) {
  return <S.Link {...restProps}>{children}</S.Link>;
};

Footer.Title = function FooterTitle({ children, ...restProps }) {
  return <S.Title {...restProps}>{children}</S.Title>;
};

Footer.Text = function FooterText({ children, ...restProps }) {
  return <S.Text {...restProps}>{children}</S.Text>;
};

Footer.Break = function FooterBreak({ children, ...restProps }) {
  return <S.Break {...restProps}>{children}</S.Break>;
};
