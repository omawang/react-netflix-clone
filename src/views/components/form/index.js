import React from 'react';

import * as S from './styles/form';

export default function Form({ children, ...restProps }) {
  return <S.Container {...restProps}>{children}</S.Container>;
}

Form.Base = function FormBase({ children, ...restProps }) {
  return <S.Base {...restProps}>{children}</S.Base>;
};

Form.Title = function FormTitle({ children, ...restProps }) {
  return <S.Title {...restProps}>{children}</S.Title>;
};

Form.Error = function FormError({ children, ...restProps }) {
  return <S.Error {...restProps}>{children}</S.Error>;
};

Form.Text = function FormText({ children, ...restProps }) {
  return <S.Text {...restProps}>{children}</S.Text>;
};

Form.TextSmall = function FormTextSmall({ children, ...restProps }) {
  return <S.TextSmall {...restProps}>{children}</S.TextSmall>;
};

Form.Link = function FormLink({ children, ...restProps }) {
  return <S.Link {...restProps}>{children}</S.Link>;
};

Form.Input = function FormInput({ ...restProps }) {
  return <S.Input {...restProps} />;
};

Form.Submit = function FormSubmit({ children, ...restProps }) {
  return <S.Submit {...restProps}>{children}</S.Submit>;
};
