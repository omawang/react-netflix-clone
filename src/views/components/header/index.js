import React, { useState } from 'react';
import { Link as ReactRouterLink } from 'react-router-dom';
import * as S from './styles/header';

export default function Header({ bg = true, src, children, ...restProps }) {
  return bg ? (
    <S.Background src={src} {...restProps}>
      {children}
    </S.Background>
  ) : (
    children
  );
}

Header.Frame = function HeaderFrame({ children, ...restProps }) {
  return <S.Container {...restProps}>{children}</S.Container>;
};

Header.Group = function HeaderGroup({ children, ...restProps }) {
  return <S.Group {...restProps}>{children}</S.Group>;
};

Header.Search = function HeaderSearch({
  searchTerm,
  setSearchTerm,
  children,
  ...restProps
}) {
  const [active, setActive] = useState(false);
  return (
    <S.Search {...restProps}>
      <S.SearchIcon
        onClick={() => setActive((active) => !active)}
        data-testid="search-click"
      >
        <img src="/images/icons/search.png" alt="Search" />
      </S.SearchIcon>
      <S.SearchInput
        value={searchTerm}
        onChange={({ target }) => setSearchTerm(target.value)}
        placeholder="Search films and series"
        active={active}
        data-testid="search-input"
      />
    </S.Search>
  );
};

Header.Logo = function HeaderLogo({ to, src, ...restProps }) {
  return (
    <ReactRouterLink to={to}>
      <S.Logo src={src} data-testid="header-bg" {...restProps} />
    </ReactRouterLink>
  );
};

Header.ButtonLink = function HeaderButtonLink({ children, ...restProps }) {
  return <S.ButtonLink {...restProps}>{children}</S.ButtonLink>;
};

Header.TextLink = function HeaderTextLink({ children, ...restProps }) {
  return <S.TextLink {...restProps}>{children}</S.TextLink>;
};

Header.Profile = function HeaderProfile({ children, ...restProps }) {
  return <S.Profile {...restProps}>{children}</S.Profile>;
};

Header.Dropdown = function HeaderDropdown({ children, ...restProps }) {
  return <S.Dropdown {...restProps}>{children}</S.Dropdown>;
};

Header.Picture = function HeaderPicture({ src, ...restProps }) {
  return <S.Picture src={`/images/users/${src}.png`} {...restProps} />;
};

Header.Feature = function HeaderFeature({ children, ...restProps }) {
  return <S.Feature {...restProps}>{children}</S.Feature>;
};

Header.FeatureCallOut = function HeaderFeatureCallOut({
  children,
  ...restProps
}) {
  return <S.FeatureCallOut {...restProps}>{children}</S.FeatureCallOut>;
};

Header.Text = function HeaderText({ children, ...restProps }) {
  return <S.Text {...restProps}>{children}</S.Text>;
};

Header.PlayButton = function HeaderPlayButton({ children, ...restProps }) {
  return <S.PlayButton {...restProps}>{children}</S.PlayButton>;
};
