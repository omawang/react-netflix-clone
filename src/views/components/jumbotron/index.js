import React from 'react';
import * as S from './styles/jumbotron';

export default function Jumbotron({
  children,
  direction = 'row',
  ...restProps
}) {
  return (
    <S.Item {...restProps}>
      <S.Inner direction={direction}>{children}</S.Inner>
    </S.Item>
  );
}

Jumbotron.Container = function JumbotronContainer({ children, ...restProps }) {
  return <S.Container {...restProps}>{children}</S.Container>;
};

Jumbotron.Pane = function JumbotronPane({ children, ...restProps }) {
  return <S.Pane {...restProps}>{children}</S.Pane>;
};

Jumbotron.Title = function JumbotronTitle({ children, ...restProps }) {
  return <S.Title {...restProps}>{children}</S.Title>;
};

Jumbotron.SubTitle = function JumbotronSubTitle({ children, ...restProps }) {
  return <S.SubTitle {...restProps}>{children}</S.SubTitle>;
};

Jumbotron.Image = function JumbotronImage({ ...restProps }) {
  return <S.Image {...restProps} />;
};
