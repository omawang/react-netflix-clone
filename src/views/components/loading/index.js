import React from 'react';

import * as S from './styles/loading';

export default function Loading({ src, ...restProps }) {
  return (
    <S.Spinner {...restProps}>
      <S.LockBody />
      <S.Picture
        src={`/images/users/${src}.png`}
        data-testid="loading-picture"
      />
    </S.Spinner>
  );
}

Loading.ReleaseBody = function LoadingReleaseBody() {
  return <S.ReleaseBody />;
};
