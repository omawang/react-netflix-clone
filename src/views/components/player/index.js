import React, { createContext, useContext, useState } from 'react';
import ReactDOM from 'react-dom';

import * as S from './styles/player';

export const PlayerContext = createContext();

export default function Player({ children, ...restProps }) {
  const [showPlayer, setShowPlayer] = useState(false);

  return (
    <PlayerContext.Provider value={{ showPlayer, setShowPlayer }}>
      <S.Container {...restProps}>{children}</S.Container>
    </PlayerContext.Provider>
  );
}

Player.Video = function PlayerVideo({ src, children, ...restProps }) {
  const { showPlayer, setShowPlayer } = useContext(PlayerContext);

  return showPlayer
    ? ReactDOM.createPortal(
        <S.Overlay
          onClick={() => setShowPlayer(false)}
          data-testid="player"
          {...restProps}
        >
          <S.Inner>
            <video id="netflix-player" controls>
              <source src={src} type="video/mp4" />
            </video>
          </S.Inner>
        </S.Overlay>,
        document.body
      )
    : null;
};

Player.Button = function PlayerButton({ children, ...restProps }) {
  const { setShowPlayer } = useContext(PlayerContext);

  return (
    <S.Button
      onClick={() => setShowPlayer((showPlayer) => !showPlayer)}
      {...restProps}
    >
      {children}
    </S.Button>
  );
};
