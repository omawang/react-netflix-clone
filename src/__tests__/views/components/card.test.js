import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { Card, Player } from '../../../views/components';

const category = 'films';
const slideRows = [
  {
    title: 'Documentaries',
    data: [
      {
        id: 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx',
        docId: 'IsdnmsdfnlkJJUds',
        title: 'Tiger King',
        description: 'Tiger King descriptions',
        genre: 'documentaries',
        maturity: '18',
        slug: 'tiger-king',
      },
    ],
  },
  {
    title: 'Comedies',
    data: [
      {
        id: 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxsss',
        docId: 'IsdnmsdfnlkJJUaa',
        title: 'The Office',
        description: 'The Office descriptions',
        genre: 'comedies',
        maturity: '11',
        slug: 'the-office',
      },
    ],
  },
];

describe('<Card />', () => {
  it('renders the <Card /> with populated data', () => {
    const { container, getByText } = render(
      <Card.Group>
        {slideRows.map((slideItem) => (
          <Card key={`${category}-${slideItem.title.toLowerCase()}`}>
            <Card.Title>{slideItem.title}</Card.Title>
            <Card.Entities>
              {slideItem.data.map((item) => (
                <Card.Item key={item.docId} item={item}>
                  <Card.Image
                    src={`/images/${category}/${item.genre}/${item.slug}/small.jpg`}
                  />
                  <Card.Meta>
                    <Card.SubTitle>{item.title}</Card.SubTitle>
                    <Card.Text>{item.description}</Card.Text>
                  </Card.Meta>
                </Card.Item>
              ))}
            </Card.Entities>
            <Card.Feature category={category}>
              <Player>
                <Player.Button>Play</Player.Button>
                <Player.Video src="/videos/bunny.mp4" />
              </Player>
            </Card.Feature>
          </Card>
        ))}
      </Card.Group>
    );

    expect(getByText('Documentaries')).toBeTruthy();
    expect(getByText('Tiger King')).toBeTruthy();
    expect(getByText('Tiger King descriptions')).toBeTruthy();

    expect(getByText('Comedies')).toBeTruthy();
    expect(getByText('The Office')).toBeTruthy();
    expect(getByText('The Office descriptions')).toBeTruthy();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the <Card /> and toggles the card feature', () => {
    const { container, queryByText, getByTestId, getByAltText } = render(
      <Card.Group>
        {slideRows.map((slideItem) => (
          <Card key={`${category}-${slideItem.title.toLowerCase()}`}>
            <Card.Title>{slideItem.title}</Card.Title>
            <Card.Entities>
              {slideItem.data.map((item) => (
                <Card.Item
                  key={item.docId}
                  item={item}
                  data-testid={`${item.slug}-item-feature`}
                >
                  <Card.Image
                    src={`/images/${category}/${item.genre}/${item.slug}/small.jpg`}
                  />
                  <Card.Meta>
                    <Card.SubTitle>{item.title}</Card.SubTitle>
                    <Card.Text>{item.description}</Card.Text>
                  </Card.Meta>
                </Card.Item>
              ))}
            </Card.Entities>
            <Card.Feature category={category}>
              <Player>
                <Player.Button>Play</Player.Button>
                <Player.Video src="/videos/bunny.mp4" />
              </Player>
            </Card.Feature>
          </Card>
        ))}
      </Card.Group>
    );

    expect(queryByText('18')).toBeFalsy();
    fireEvent.click(getByTestId('tiger-king-item-feature'));
    expect(queryByText('18')).toBeTruthy();

    fireEvent.click(getByAltText('Close'));
    expect(queryByText('18')).toBeFalsy();

    expect(queryByText('PG')).toBeFalsy();
    fireEvent.click(getByTestId('the-office-item-feature'));
    expect(queryByText('PG')).toBeTruthy();

    fireEvent.click(getByAltText('Close'));
    expect(queryByText('PG')).toBeFalsy();

    expect(container.firstChild).toMatchSnapshot();
  });
});
