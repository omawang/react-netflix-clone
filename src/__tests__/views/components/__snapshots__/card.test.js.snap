// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> renders the <Card /> and toggles the card feature 1`] = `
<div
  class="card__Group-sc-11y44z9-2 tanHN"
>
  <div
    class="card__Container-sc-11y44z9-1 dCcCtS"
  >
    <p
      class="card__Title-sc-11y44z9-0 kjskVN"
    >
      Documentaries
    </p>
    <div
      class="card__Entities-sc-11y44z9-3 kygyag"
    >
      <div
        class="card__Item-sc-11y44z9-7 lbtzkX"
        data-testid="tiger-king-item-feature"
      >
        <img
          class="card__Image-sc-11y44z9-8 dZirdd"
          src="/images/films/documentaries/tiger-king/small.jpg"
        />
        <div
          class="card__Meta-sc-11y44z9-4 hKYdNM"
        >
          <p
            class="card__SubTitle-sc-11y44z9-5 gGBOEJ"
          >
            Tiger King
          </p>
          <p
            class="card__Text-sc-11y44z9-6 bptwhU"
          >
            Tiger King descriptions
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card__Container-sc-11y44z9-1 dCcCtS"
  >
    <p
      class="card__Title-sc-11y44z9-0 kjskVN"
    >
      Comedies
    </p>
    <div
      class="card__Entities-sc-11y44z9-3 kygyag"
    >
      <div
        class="card__Item-sc-11y44z9-7 lbtzkX"
        data-testid="the-office-item-feature"
      >
        <img
          class="card__Image-sc-11y44z9-8 dZirdd"
          src="/images/films/comedies/the-office/small.jpg"
        />
        <div
          class="card__Meta-sc-11y44z9-4 hKYdNM"
        >
          <p
            class="card__SubTitle-sc-11y44z9-5 gGBOEJ"
          >
            The Office
          </p>
          <p
            class="card__Text-sc-11y44z9-6 bptwhU"
          >
            The Office descriptions
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Card /> renders the <Card /> with populated data 1`] = `
<div
  class="card__Group-sc-11y44z9-2 tanHN"
>
  <div
    class="card__Container-sc-11y44z9-1 dCcCtS"
  >
    <p
      class="card__Title-sc-11y44z9-0 kjskVN"
    >
      Documentaries
    </p>
    <div
      class="card__Entities-sc-11y44z9-3 kygyag"
    >
      <div
        class="card__Item-sc-11y44z9-7 lbtzkX"
      >
        <img
          class="card__Image-sc-11y44z9-8 dZirdd"
          src="/images/films/documentaries/tiger-king/small.jpg"
        />
        <div
          class="card__Meta-sc-11y44z9-4 hKYdNM"
        >
          <p
            class="card__SubTitle-sc-11y44z9-5 gGBOEJ"
          >
            Tiger King
          </p>
          <p
            class="card__Text-sc-11y44z9-6 bptwhU"
          >
            Tiger King descriptions
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card__Container-sc-11y44z9-1 dCcCtS"
  >
    <p
      class="card__Title-sc-11y44z9-0 kjskVN"
    >
      Comedies
    </p>
    <div
      class="card__Entities-sc-11y44z9-3 kygyag"
    >
      <div
        class="card__Item-sc-11y44z9-7 lbtzkX"
      >
        <img
          class="card__Image-sc-11y44z9-8 dZirdd"
          src="/images/films/comedies/the-office/small.jpg"
        />
        <div
          class="card__Meta-sc-11y44z9-4 hKYdNM"
        >
          <p
            class="card__SubTitle-sc-11y44z9-5 gGBOEJ"
          >
            The Office
          </p>
          <p
            class="card__Text-sc-11y44z9-6 bptwhU"
          >
            The Office descriptions
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
